Optimize thread pool implementation:

nbougalis opened the pull request:
The existing thread pool code uses several layers of indirection which uses a custom lock-free stack, and offers functionality that supports features that are never used (e.g. the ability to dynamically adjust the number of threads in the pool).

This refactoring aims to simplify the code, making it easier to reason about (although lock-free multi-threaded code is always tricky) what is happening, and reduce the latency of the thread pool internals.

<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

scottschurr reviewed (changes requested):
First, thanks for simplifying this unnecessarily complicated part of the codebase.  Several years back I spent some time fixing a corner case when the number of threads in `Workers` decreased.  That bug, and the time spent fixing and testing it, was pointless since the feature was gratuitous.  Less complexity means fewer bugs.  So this pull request is a win.

However, as it stands, this pull request makes the full suite of unit tests take more than twice as long to execute.  Not good.  It's fixable however.

I have a number of suggestions for what I think are improvements for `Workers`.  Most of them are naming and tidiness.  But using `std::thread::join()` rather than `detach()`is critical to not losing the shutdown performance.  You can see all of those accumulated `Workers` changes here: https://github.com/scottschurr/rippled/commit/0c345a02f88b234c38523dfe1a3bec125f75cb15

There are also some less important suggestions I made for `JobQueue`.  You can see those implemented here: https://github.com/scottschurr/rippled/commit/9ce534745035a4cbe96132631eba63b64a076401

Feel free to cherry-pick those commits if you would like.

nbougalis commented on src/test/core/Workers_test.cpp:
```diff
@@ -31,149 +31,4 @@ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
-
-BEAST_DEFINE_TESTSUITE(Workers, core, ripple);
-
 }  // namespace ripple
```
Agreed, my bad. I guess I could rework the unit test to actually exercise the new code, but I don't know that it really matters given how the complex functionality (increasing/decreasing thread count) is now a thing of the past.

nbougalis commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,113 @@
-    lk.unlock();
+                while (!stopping_)
+                {
+                    auto t = tail_.load();
```
`th` and `ta` it is :)

nbougalis commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,113 @@
-    ++wakeCount_;
-    wakeup_.notify_one();
+    while (threads_.load() != 0)
+        std::this_thread::sleep_for(std::chrono::seconds(1));
```
Wowza... OK, I'll look at your proposed changes and rethink this.

nbougalis commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,113 @@
+                    beast::setCurrentThreadName(n + " [zZz]");
+
+                    // If there's nothing to do, go to sleep
+                    while (h == t && !stopping_.load(std::memory_order_relaxed))
```
I think that in this case it's safe, but I hear what you're saying. I'm ok going with the default `memory_order` here, even though it's not _strictly_ needed.

seelabs replied:
I agree with both @scottschurr and @nbougalis: I think the code is OK as-is, but we should default to `memory_order_sequential` unless there's a compelling reason not to.

nbougalis commented:
So having looked at your proposed changes, I'm not sure I'm sold. I dislike the need for a `vector` to hold the `thread` instances. I suspect that simply changing the `std::this_thread::sleep_for(std::chrono::seconds(1));` to something like `std::this_thread::yield()` will be sufficient to restore the performance.

scottschurr commented:
One person's cake is another person's poison.  Indeed, another approach would be to see if a `yield()` instead of a `sleep_for()` returns the performance.  But I _like_ the vector of threads because it leaves an artifact behind that future maintainers can easily see.  A thread that is `detach`ed is virtually invisible to a maintainer.  A `join()` on shutdown is ubiquitous in the code base, effective, and easy for a maintainer to understand.

nbougalis commented:
> One person's cake is another person's poison. Indeed, another approach would be to see if a `yield()` instead of a `sleep_for()` returns the performance. But I _like_ the vector of threads because it leaves an artifact behind that future maintainers can easily see. A thread that is `detach`ed is virtually invisible to a maintainer. A `join()` on shutdown is ubiquitous in the code base, effective, and easy for a maintainer to understand.

So, `yield` does return the performance back to where it was. And while I _personally_ prefer not having the `std::vector` to be joined (especially since C++20 brings with it the ability to `wait` on `std::atomic`) I see your point: the `std::vector` does expose the "artifact" for future maintainers, which is something I should be mindful of.

I'll think about this some more.

seelabs commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,113 @@
+                threads_--;
+            },
+            count);
+        t.detach();
```
I'd probably keep a collection of threads in the object rather than `detach` here.

nbougalis replied:
@scottschurr, @seelabs: I thought about this and actually wrote the version with the `std::vector<std::thread>` and _something_ about it rubbed me the wrong way. I guess it's because actually holding onto the handle doesn't really serve a purpose.

Unless you're vetoing the current version (and if you feel strongly about it, please do!), I'd prefer to stick with the version as is.

seelabs reviewed (commented):
Very nice code cleanups. We need to fix the slow-downs, but from a code complexity POV this is a nice win.

I have a concern over changing the behavior of uncaught exceptions; I'm not totally against the change, but I don't think it's a small change either.

nbougalis commented:
@seelabs writes:

>This is a large behavior change that probably should get it's own patch and not be folded into an unrelated change. There are times where we purposely crash the app on unhandled exceptions - because we don't want to continue in an insane state. We've discussed making changes like this in the past, and have always opted to keep the same behavior.
>
>I'm not necessarily against making this change, but at minimum we need to make this change much more visible.

I hear what you're saying; is this behavior change appropriate for this PR?

Let's take a step back. My position is that relying on this "hidden" behavior (i.e. that the job queue doesn't catch exceptions and that to `throw` signals "ABANDON SHIP!") is anathema to good software engineering perspective. We've both encountered core dumps that _might_ be traceable to this.

At some point we're going to need to fix this.

So, the way I see it is that:

- Either the job handlers need to be marked as `noexcept`, so that the compiler knows they won't throw and can warn if they do and _if_ something slips through, the runtime will kill the process through [`std::terminate`](https://en.cppreference.com/w/cpp/error/terminate); or
- The job queue must catch these exceptions, warn loudly, and, potentially terminate.

I think that terminating is reasonable because, as you said, it's hard for the job queue to properly handle exceptions. So I would suggest that we leave the new `try`/`catch` in place and simply add an explicit call to `LogicError` (which, honestly, is horribly named and ought to be renamed to `ripple::abort(std::string const& msg)` and should just call `std::abort`, but that's another story) in the `catch` handler.

nbougalis commented on src/ripple/core/JobTypeData.h:
```diff
@@ -8,7 +8,7 @@
 
     THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
-    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    WITH  REGARD  TO  THIS  SOFTWARE  INCPerLUDING  ALL  IMPLIED  WARRANTIES  OF
```
What happened here? :thinking: I'll fix this.

scottschurr commented:
Just FYI, the Refactor JobQueue commit message has some junk tagged onto the end.  You may wish to clean that up.

scottschurr reviewed (commented):
Identified a few more changes for you to consider.

FWIW, I timed unit tests with your change which uses `std::detach` and `std::yield`.  I also timed unit tests with an implementation that replaces those with `std::join`.  The unit test timings came out comparable.  So, although I think the `std::join` is more maintainable, I'm okay with your approach using `std::detach` and `std::yield`.

HowardHinnant commented on src/ripple/core/impl/JobQueue.cpp:
```diff
@@ -299,72 +248,92 @@ JobQueue::isStopped() const
+
+        LogicError(
+            beast::getCurrentThreadName() +
+            ": Uncaught exception handler invoked with no exception_ptr");
```
It is true that catching an exception erases the stack trace information pointing to where the exception was thrown.  If unique, the error message in `what()` might pinpoint that the place of throw, but that still doesn't recover the stack trace down to that line of code.

thejohnfreeman replied:
I'm assuming we're talking about the exception pointed by `eptr`. For this function to be called, that exception must have already been caught to be captured in `eptr`. If what you're saying is true, then the stack trace is already lost when this function is entered. Are we sure that is true? Or maybe I'm misunderstanding.

nbougalis commented on src/ripple/core/JobTypes.h:
```diff
@@ -21,171 +21,147 @@
-    size() const
+    JobType const type;
+
+    /// The job's display name (when we get C++20 this could be std::string)
+    char const* const name;
```
I think I'm OK with that, although `std::string_view` is at least 8 bytes larger than a naked pointer. I wonder if there's enough "padding" in the layout to absorb this.

Not that I care about 8 bytes in _this_ case, since even with a 1,000,000 jobs in the queue (an insanely and unfathomably large number of 'em) we'd only be "wasting" 8MB.

nbougalis commented on src/ripple/core/JobQueue.h:
```diff
@@ -242,36 +241,34 @@ class JobQueue : private Workers::Callback
 
-    beast::Journal m_journal;
+    beast::Journal journal_;
     mutable std::mutex m_mutex;
```
I think I just missed it. I can look into getting it done.

nbougalis commented on src/ripple/core/JobTypes.h:
```diff
@@ -21,171 +21,147 @@
+     */
+    int const limit;
+
+    /** Average and peak latencies for this job type. 0 is none specified */
```
Good point.

nbougalis commented on src/ripple/core/impl/JobQueue.cpp:
```diff
@@ -80,39 +71,28 @@ JobQueue::addRefCountedJob(
-        m_workers.getNumberOfThreads() > 0);
+        workers_.count() > 0);
+
+    auto& data = jobData_[type];
```
Let me look at this more closely.

> It's worth remembering that not everyone who's working on this code base is as clever as you are.

Most are cleverer, and that's the way I like it. :grinning:

nbougalis commented on src/ripple/core/impl/JobQueue.cpp:
```diff
@@ -299,72 +248,94 @@ JobQueue::isStopped() const
-void
-JobQueue::getNextJob(Job& job)
+Job
+JobQueue::getNextJob()
```
Actually, when I was doing this refactor, I wanted to eliminate this function and turn it into a direct-invoked lambda, precisely because of the issue that you describe; it would prevent any possibility of abuse. I think I may do that.

nbougalis commented on src/ripple/core/impl/LoadEvent.cpp:
```diff
@@ -18,53 +18,69 @@
+                                       std::chrono::steady_clock::duration,
+                                       std::chrono::steady_clock::duration) {
+    /* do nothing */
+};
```
Good point. I'm fine with that.

nbougalis replied:
Effectively resolved by making `callback_` a `std::reference_wrapper<const>`.

nbougalis commented on src/ripple/core/JobTypes.h:
```diff
@@ -21,171 +21,147 @@
-    const_iterator
-    begin() const
+    bool
+    special() const
```
Yeah, I hear you. This is an unfortunate consequence of using the job queue to keep track of performnace and execution of certain tasks, as opposed to... you know... queueing jobs.

I'm fine with renaming, but (imo) that's a workaround; the right thing to do is to use one class to queue jobs and dispatch them for execution for and another to track the timing of operations.

nbougalis commented on src/ripple/core/impl/LoadEvent.cpp:
```diff
@@ -18,53 +18,69 @@
+    if (this != &other)
+    {
+        name_ = std::move(other.name_);
+        callback_ = other.callback_;
```
I think I've resolved this effectively.

greg7mdp commented on src/ripple/core/impl/LoadEvent.cpp:
```diff
@@ -18,58 +18,71 @@
+    {
+        name_ = std::move(other.name_);
+        callback_ = other.callback_;
+        running_ = other.running_;
```
should we also set:
`neutered_ = other.neutered_;`?

nbougalis replied:
Hmm... catching the case where we assign an already neutered object? Yeah, I think it's best if we do that.

greg7mdp commented on src/ripple/core/impl/LoadEvent.cpp:
```diff
@@ -18,58 +18,71 @@
+    if (this != &other)
+    {
+        name_ = std::move(other.name_);
+        callback_ = other.callback_;
```
should probably be:
`callback_ = std::move(other.callback_);`

nbougalis replied:
I don't think `std::reference_wrapper` is move-assignable?

greg7mdp replied:
Yes, you are correct, sorry I thought it stored the `std::function`, not a `std::reference_wrapper` to it.

nbougalis commented on src/ripple/core/LoadMonitor.h:
```diff
@@ -83,6 +87,12 @@ class LoadMonitor
     std::chrono::milliseconds mTargetLatencyPk;
```
Done.

scottschurr commented:
I tried out the most recent version of these changes on macOS.  I see two things that need to be addressed:

1. Clang pointed out that `JobTypeData(JobTypeData&& other) = default` is invalid.  Since `LoadMonitor` is immovable, then `JobTypeData` must also be immovable.  This line could be removed or you could use `= delete` instead.

2. While running the unit tests on a debug build, the `Coroutine` test failed with a segmentation fault.  Here is the message I got:
```
ripple.core.Coroutine thread specific storage
Logic error: Job: LocalValue-Test: Exception caught during task processing: mutex lock failed: Invalid argument
Segmentation fault: 11
```
The stack trace points at JobQueue.cpp line 265: https://github.com/XRPLF/rippled/pull/4172/files#diff-85e882e68cc55760882821eba100a325cad49b5bce66e21047bd861b3a7e6fedR262-R264

The good news is that the exception message got out.  The bad news is that since we caught the exception the trace to the actual source of the exception is lost.

The `Coroutine` unit test crash is not consistent, but if I loop the unit tests I get a crash every 20 repetitions or so.  So it's not hard to reproduce.

I removed the `try/catch` in `Workers.cpp` to see what I could find out from the stack trace.  Here's that stack trace from the crashed thread:

```
Thread 6 Crashed:: Job: LocalValue-Test
0   libsystem_kernel.dylib        	0x00007fff6996e33a __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff69a2ae60 pthread_kill + 430
2   libsystem_c.dylib             	0x00007fff698f5808 abort + 120
3   libc++abi.dylib               	0x00007fff66b54458 abort_message + 231
4   libc++abi.dylib               	0x00007fff66b458a7 demangling_terminate_handler() + 238
5   libobjc.A.dylib               	0x00007fff686805b1 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff66b53887 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff66b561a2 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 27
8   libc++abi.dylib               	0x00007fff66b56169 __cxa_throw + 113
9   libc++.1.dylib                	0x00007fff66b3055b std::__1::__throw_system_error(int, char const*) + 77
10  libc++.1.dylib                	0x00007fff66b2754d std::__1::mutex::lock() + 29
11  rippled                       	0x000000010da12e83 std::__1::lock_guard<std::__1::mutex>::lock_guard(std::__1::mutex&) + 35 (__mutex_base:91)
12  rippled                       	0x000000010da1260d std::__1::lock_guard<std::__1::mutex>::lock_guard(std::__1::mutex&) + 29 (__mutex_base:91)
13  rippled                       	0x000000010da2698a ripple::test::Coroutine_test::gate::signal() + 42
14  rippled                       	0x000000010da35bf9 ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()::operator()() const + 89 (Coroutine_test.cpp:172)
15  rippled                       	0x000000010da35b8e ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>::operator()() + 30 (ClosureCounter.h:140)
16  rippled                       	0x000000010da35b4d decltype(std::__1::forward<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>&>(fp)()) std::__1::__invoke<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>&>(ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>&) + 29 (type_traits:3545)
17  rippled                       	0x000000010da35afd void std::__1::__invoke_void_return_wrapper<void>::__call<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>&>(ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>&) + 29 (__functional_base:349)
18  rippled                       	0x000000010da35acd std::__1::__function::__alloc_func<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>, std::__1::allocator<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()> >, void ()>::operator()() + 29 (functional:1546)
19  rippled                       	0x000000010da3491e std::__1::__function::__func<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()>, std::__1::allocator<ripple::ClosureCounter<void>::Substitute<ripple::test::Coroutine_test::thread_specific_storage()::'lambda0'()> >, void ()>::operator()() + 30 (functional:1720)
20  rippled                       	0x000000010b3fb1b5 std::__1::__function::__value_func<void ()>::operator()() const + 53 (functional:1873)
21  rippled                       	0x000000010b3fb155 std::__1::function<void ()>::operator()() const + 21 (functional:2548)
22  rippled                       	0x000000010bd19b7d ripple::Job::doJob() + 141 (Job.cpp:68)
23  rippled                       	0x000000010bd20593 ripple::JobQueue::processTask(unsigned int) + 723 (JobQueue.cpp:328)
24  rippled                       	0x000000010bd64341 ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0::operator()(unsigned int) const + 977 (Workers.cpp:80)
25  rippled                       	0x000000010bd63ee1 decltype(std::__1::forward<ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0>(fp)(std::__1::forward<unsigned int>(fp0))) std::__1::__invoke<ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0, unsigned int>(ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0&&, unsigned int&&) + 49 (type_traits:3545)
26  rippled                       	0x000000010bd63e2e void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0, unsigned int, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0, unsigned int>&, std::__1::__tuple_indices<2ul>) + 62 (thread:274)
27  rippled                       	0x000000010bd63576 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ripple::Workers::Workers(ripple::Workers::Callback&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int)::$_0, unsigned int> >(void*) + 118 (thread:284)
28  libsystem_pthread.dylib       	0x00007fff69a2b109 _pthread_start + 148
29  libsystem_pthread.dylib       	0x00007fff69a26b8b thread_start + 15
```

nbougalis commented:
The issue that @scottschurr encountered had to do with timing, but it did lead me to discover an implementation issue that could cause a worker thread to sleep despite the fact that work was waiting.

I've reworked the code to address this and now leverage more C++20 functionality, since the PR to enable this was merged into *`develop`*.

I know it's annoying to have to do this, but please re-review the first commit (that changes the `Workers` logic) in its entirety.

scottschurr commented:
With the most recent code I consistently get the following failure on the JobQueue unit test:
```
$ build/rippled --unittest=JobQueue
ripple.core.JobQueue
Assertion failed: (!stopped_), function postCoro, file JobQueue.h, line 391.
Abort trap: 6
```
I'm running macOS 12.5.1, Apple clang version 13.1.6 (clang-1316.0.21.2.5).

nbougalis commented:
> With the most recent code I consistently get the following failure on the JobQueue unit test:
> 
> ```
> $ build/rippled --unittest=JobQueue
> ripple.core.JobQueue
> Assertion failed: (!stopped_), function postCoro, file JobQueue.h, line 391.
> Abort trap: 6
> ```
> 
> I'm running macOS 12.5.1, Apple clang version 13.1.6 (clang-1316.0.21.2.5).

Yeah, that was just sloppy of me. It's been fixed and the ugly-looking code in the unit tests has been removed.

chennakeshava1998 commented on src/ripple/core/JobQueue.h:
```diff
@@ -237,41 +242,47 @@ class JobQueue : private Workers::Callback
 
     // The number of jobs currently in processTask()
-    int m_processCount;
+    int activeThreads_ = 0;
```
I'm trying to understand the semantic meaning of the variable names like activeThreads_, coroutines and workers. In parallel computing, these terms are used in a loosely inter-changeable fashion. There are a lot of comments about the different methods of the Coro class, but can we give a definition of these terms too? 

My understanding is that: Coroutines (objects of the type Coro) are function pointers which are queued for execution in order of their respective priorities. (I know that this is not an accurate description of Coro class, but something like this would be very helpful).

Also, I couldn't find the usages of activeThreads_ data member. Is it used in TaskQueue::processTask() ?

chennakeshava1998 commented on src/ripple/core/JobQueue.h:
```diff
@@ -334,7 +334,14 @@ class JobQueue : private Workers::Callback
     // Invariants:
     //  <none>
```
Previously, what did the `m_` prefeix refer to? Is it member data?

scottschurr reviewed (changes requested):
I left a number of comments, many of which are nits.  However I think I spotted a couple of thread race conditions that should be fixed.  There's also the `for` loop in Coroutine_test.cpp that runs `thread_specific_storage` 1000 times.  I'm pretty sure we don't want that to happen every time anyone runs the unit tests.

nbougalis commented on src/test/core/Coroutine_test.cpp:
```diff
@@ -179,7 +187,9 @@ class Coroutine_test : public beast::unit_test::suite
         incorrect_order();
-        thread_specific_storage();
+
+        for (int i = 0; i != 1000; ++i)
+            thread_specific_storage();
```
Oops, yeah, that's not meant to be there.

nbougalis commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,121 @@
+
+                        // Exceptions should never bubble up to here but
+                        // just in case, if one does catch it.
+                        try
+                        {
+                            callback_.processTask(instance);
+                        }
+                        catch (...)
+                        {
+                            callback_.uncaughtException(
+                                instance, std::current_exception());
+                        }
```
I can get behind this. In theory, nothing in the job queue should _ever_ throw, and if it does, it's a logic error. The idea was to allow the "driver" behind the `Workers` class to try and report this appropriately.

I'll rethink this.

nbougalis commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,121 @@
-        {
-            // See if we can reuse a paused worker
-            Worker* worker = m_paused.pop_front();
+                ++threads_;
```
Good point.

nbougalis commented on src/ripple/core/Job.h:
```diff
@@ -98,37 +93,34 @@ class Job : public CountedObject<Job>
+        the description may include additional information or context that
+        distinguishes this from other jobs of the same type.
+     */
+    [[maybe_unused]] std::string const&
```
The `[[maybe_unused]]` is intended to bind to the function (which, for example, may only ever be called for debug builds); it shouldn't bind to the return value.

scottschurr replied:
Yes, you're right, `[[maybe_unused]]` binds to the method, not to the return value.  I was mixing up the behavior with `[[nodiscard]]`.  Sorry.

nbougalis commented on src/ripple/core/Job.h:
```diff
@@ -142,12 +134,11 @@ class Job : public CountedObject<Job>
     operator>=(const Job& j) const;
```
I thought about this, but this PR was already doing a lot and I figured I could wait. But I'm fine with adding it if you're fine with it being added.

scottschurr replied:
It certainly can wait.  But I'm also fine with adding it now.  I think it's low risk and reduces some effectively redundant code.  Your call.

nbougalis commented on src/ripple/core/LoadEvent.h:
```diff
@@ -20,42 +20,36 @@
+#include <array>
 #include <chrono>
-#include <memory>
+#include <functional>
```
We need it for `std::reference_wrapper` but I guess we get that from `LoadMonitor.h`

scottschurr replied:
Oh, good point, I wasn't thinking about `std::reference_wrapper` when I made that comment.  Sorry.  I'm fine with either leaving it or having it arrive sideways from `LoadMonitor.h`.  Your call.

scottschurr commented:
The most recent version of this pull request fails the clang-format check.  The top-most commit on this brand will fix that: https://github.com/scottschurr/rippled/commits/nik-jqt-clang-format

Other than the clang-format problems, is this branch ready to be re-reviewed?  Or is there more work to be done?

nbougalis commented:
@scottschurr Thanks for the patience, the comments and the `clang-format` fix. I believe I've addressed everything, rebased the code to the latest and it should be ready for review.

I know there's a couple of `clang-format` failures still; I'll work around them. The existing formatting rules don't like attributes very much (e.g. `[[likely]]`) which is unfortunate.

nbougalis commented:
Ping...

intelliot commented:
@thejohnfreeman @seelabs @scottschurr - looks like this is ready for (re)review.

If you won't be able to have an initial look at this within 2 weeks (and provide some feedback), please comment here to let us know.

scottschurr commented:
@intelliot, I'm not ignoring you, but I have another higher priority code review I'm working on.  I can't predict whether or not I'll get to this one in the coming two weeks.

nbougalis commented:
This isn't a high priority item, imo. I'm sure Scott et al are focused on business priorities.

scottschurr reviewed (commented):
I should start out by saying that I'm really sorry that this pull request has been sitting stagnant for so long.  It's hard to have code that you've put effort into just sit and be ignored for such a long time.

At a request from @intelliot I moved this pull request to the top of my review queue.  I was requested to only verify that the changes I had flagged had been made.  So I did that.  But even with that cursory look I stumbled across a place that sure looks like it is missing a mutex lock.

So I can say that it looks like all of the important review comments I had made earlier have been addressed.  But the most recent rework contains enough significant changes that it probably should have a full review.  I'm fine with either me or someone else doing that additional review.  But I'm not ready to give this code a thumbs up yet.

nbougalis commented:
@scottschurr, no worries. I know this was (a) a large PR; (b) not a priority; and (c) you're swamped. I appreciate you taking another look.

Re: your comment about an apparently missing lock in this bit of code:

```
return {jobData_[t].load().sample(), std::move(name), true};
```

You say:

>Pretty much all of the other accesses to `jobData_` are done with `mutex_` locked. I have the uneasy sensation that there's a lock missing here.

Here's my (preliminary) rationale for why a lock isn't missing:

The `jobData_` array is constructed once and each entry in the array contains a `LoadMonitor` instance load_. The `load()` member function returns a reference to the given member.

The `LoadMonitor` instance contains a `LoadSampler` member variable which is:

```
using LoadSampler = std::function<void(
    char const*,
    std::chrono::steady_clock::duration,
    std::chrono::steady_clock::duration)>;
```

And `sample()` returns a `std::reference_wrapper<LoadSampler const>` to that member.

Given that, I _think_ that it's fine as is, although it's been a long time since I've even looked at this code and I will have to re-check to be sure.

I'll take a closer looks when I rebase this.

scottschurr commented on src/ripple/core/impl/JobQueue.cpp:
```diff
@@ -150,98 +134,83 @@ JobQueue::getJobCountGE(JobType t) const
-        return {};
-
-    return std::make_unique<LoadEvent>(iter->second.load(), name, true);
+    return {jobData_[t].load().sample(), std::move(name), true};
```
I spent some more time spelunking through the code today.  As far as I can tell it's okay that there is no lock here.  Sorry for the false alarm.  My instinct tells me that it would be good to leave a comment for why that lock is not needed.  But that's not essential.

In my defense, I think it's hard to review code using a type that requires locking only some of the time.  But I also know that part of the design predates your modifications, @nbougalis.

scottschurr reviewed (approved):
I dug into the details regarding the comment I left in a previous review.  It looks like my suspicions were unfounded.  As far as I can tell no lock is needed at that location.

Since I don't anticipate spending more time on this review, and I don't want to be responsible for holding up this pull request any longer, I'm giving the thumbs up.

The removal of `beast::LockFreeStack` and `semaphore`, `constexpr`-izing `jobTypes`, and the massive simplification of `Workers` are all good steps forward for this code base.  Thanks for the contribution, @nbougalis.

scottschurr reviewed (approved):
:+1:

intelliot commented:
@nbougalis if this PR is ready for review and merging (in your opinion), then please avoid force-pushing to the branch. Best practices would require that any changes to the PR after a review should invalidate that review.

Also, when a branch is force-pushed or any changes are made to it, please comment on the PR to explain why the push / change was made.

nbougalis commented:
How can I avoid force-pushing if I rebase a branch to resolve merge conflicts with upstream?

intelliot commented:
@nbougalis Instead of rebasing, just do a merge commit (`git merge`). The merge commit resolves merge conflicts with upstream. The merge commit will go away when the PR is merged to `develop` because we use `Squash and merge`.

intelliot commented:
@thejohnfreeman what's the status of your review of this?

intelliot commented:
@seelabs will you be able to review this?

seelabs commented:
@intelliot I won't be able to review this week for sure

thejohnfreeman commented on src/ripple/core/impl/Workers.cpp:
```diff
@@ -17,264 +17,119 @@
-            // See if we can reuse a paused worker
-            Worker* worker = m_paused.pop_front();
+        std::thread th(
+            [this, name](unsigned int instance) {
```
Just a preference, but why capture `this` and `name` but not `instance` / `count`? Could pass them all as parameters and turn this lambda into a function pointer. No biggie, just wondering how the distinction was made.

thejohnfreeman commented on src/ripple/core/JobQueue.h:
```diff
@@ -77,54 +78,62 @@ class JobQueue : private Workers::Callback
-              without a corresponding yield.
+        /** Resume execution of a suspended coroutine on the current thread.
+
+            The coroutine will continues execution from where it last left,
```
continues -> continue

intelliot commented:
@nbougalis - does this PR still make sense to merge? If so, please consider the above comments and update the branch to be up-to-date with `develop`. If not, then share a brief explanation and close. Thanks!
